Fix multi-line block titles

rchurchley opened the pull request:
For quite a while, ᴍᴇᴛʀᴏᴘᴏʟɪs has had trouble with block environments with long titles:

![screen shot 2016-02-24 at 10 05 00 am](https://cloud.githubusercontent.com/assets/1131743/13295584/6d92ff3c-dade-11e5-89c6-a6b806650f2b.png)

This pull request solves this issue while maintaining the current behaviour for single-line blocks:
![screen shot 2016-02-24 at 10 05 55 am](https://cloud.githubusercontent.com/assets/1131743/13296399/3d5e0556-dae2-11e5-8feb-7a7d0a6de87a.png)
![screen shot 2016-02-24 at 10 06 45 am](https://cloud.githubusercontent.com/assets/1131743/13296528/d2902636-dae2-11e5-89b2-e51ffa74dc75.png)

Technical details:
- Previously, we fixed the height and depth of the `beamercolorbox` in block titles, and added padding to the right and left when `block title.bg` was nonempty. Now, we set `sep` for the `beamercolorbox` and add negative padding to the right and left when `block title.bg` is empty. This is similar to how we fixed multi-line frame titles a while back.
- Just replacing `dp` and `ht` with `sep` wouldn't be enough, though. The padding at the top of a beamer box seems to actually be `sep + parskip`; since ᴍᴇᴛʀᴏᴘᴏʟɪs increases the `\parskip` length, we'd have too much padding at the top if we didn't do something more. The solution is to reset `\parskip` to zero at the beginning of `block begin` and restore the ᴍᴇᴛʀᴏᴘᴏʟɪs `\parskip` length when the block body begins.
- This pull request also adds some wiggle room to the `rightskip` for the block title to produce a good ragged right edge.
- Finally, I added some comments explaining why we can't just use `colsep`, which would be the ideal solution if it actually worked.

benjamin-weiss commented:
Seems fine to me, although the solution is bit weird. I would probably prefer a solution the other way around. Set `sep=0pt` by default, add custom struts at the beginning and end (like I proposed [here](https://github.com/matze/mtheme/issues/165) with [this branch](https://github.com/benjamin-weiss/mtheme/commit/047c5ab4412e9863ecf40e40499a8e260b156511) for the frame titles but never submitted the PR) and set `leftskip` and `rightskip` if the background color is *not* empty.

rchurchley commented:
Coming from a web/CSS background, setting the desired padding and compensating for a special case with a negative margin was actually the more intuitive solution to me.

I prefer using Beamer's higher-level options where possible rather than rolling our own solutions with struts. Using Beamer's mechanisms mean that our bugs will at least be consistent with other themes. I'm sure that a strut-based solution is possible, but I'm not sure how well I could debug it, or to anticipate the relevant edge cases where our solution would differ from other themes. (3+ line titles? Accented capital letters?)

In an ideal world, of course, we'd just use `colsep` and not have any colour-conditional code at all. I suppose that's another advantage of the `sep` solution — if Beamer fixes the bug with `colsep` and inherited empty colours, it will be very easy to take advantage of that.

benjamin-weiss commented:
I have done my fair share of CSS too and that is the reason why the `sep` or `colsep` parameter looks so strange to me. Consider the following example in HTML/CSS:

```HTML
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Padding Test</title>
 </head>
<body>
<p style="padding: 1ex;">
<div style="padding:1ex; background:#E6E8E8; font-size: 30pt; font-family: 'Fira Sans'; display: inline; margin-left:1ex;">
ALL CAPS
</div>
<div style="padding:1ex; background:#D0D4D5; font-size: 30pt; font-family: 'Fira Sans'; display: inline">
Regular
</div>
</p>
</body>
```
You can see that whatever text you write inside the `div`, it has always the same height. That is not the case with `beamercolorbox`es. A `beamercolorbox` with the same font size and the same `sep` can have a different height. I illustrated this earlier with this image.

![1c7afd4e-2ebc-11e5-96a6-a7d19d75b1df](https://cloud.githubusercontent.com/assets/9158719/13313756/9a96b9ac-dba0-11e5-802f-7dd144f12503.png)

With CSS and a fixed padding setting the distances, a_2, c and b are the same. My custom solution does exactly this.

rchurchley commented:
Ah, I see my misunderstanding. By "the solution is bit weird", I thought you were objecting on code readability grounds. You bring up a very valid point about box heights. 

I originally thought that maintaining the height of block titles was not as important as it is for the frametitles, as blocks are usually not positioned in a way that the eye naturally compares the heights. But you're absolutely right; there are common use cases where my current solution looks bad:

![screenshot](https://cloud.githubusercontent.com/assets/1131743/13332533/448588f6-dbb7-11e5-8657-54529d06a5a9.png)

Implementing the padding entirely with struts, as you did in #165, would be one way to go. However, since we don't have to satisfy the constraints of an `allsmallcaps` option for the block titles, I don't think we need to go quite so far. It should be fine to keep `sep` for the padding inside the box, and just use `\vphantom{ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghikjlmnopqrstuvwxyz()}` to ensure a consistent line height.

![screenshot-1](https://cloud.githubusercontent.com/assets/1131743/13335213/bf45d2be-dbc4-11e5-930c-314db0cec223.png)

What do you think?

matze commented:
Hmm, although the latter is objectively more consistent the (relatively) larger title backgrounds seem to emphasize them even more and de-emphasize the actual content. On the other hand I won't use filled blocks anyway, so this is good for me. I'll wait for @benjamin-weiss' feedback before merging this in.

benjamin-weiss commented:
The screenshot is a bit small and I don't have time right now to check it myself. But I also think the block title block is a bit to big. That is one of the reasons I implemented my "overcomplicated" custom struts. Right now a_1, c and "b_2" are the same length. I will check it out thoroughly tonight (in around 10 hours).

rchurchley commented:
@matze Actually, that's another good point. Without making the struts conditional on the background, this solution also affects transparent blocks, adding what I feel is too much space between the block title and block body (see below left)

![untitled](https://cloud.githubusercontent.com/assets/1131743/13346396/d72383e4-dc1a-11e5-9171-c9341cacdf9d.png)

Unless we make our struts conditional on the block background, we'll also have to be careful to test how our solutions affect transparent blocks.

rchurchley commented:
I've tweaked the numbers a little bit in the hopes of getting something a little closer to what we'd like, based partly on Beamer's default measurements (`colsep=0.75ex`) and partly on a rough calculation that suggests parentheses contribute about `0.25ex` to each of the top and bottom padding.

![screen shot 2016-02-26 at 1 18 54 am](https://cloud.githubusercontent.com/assets/1131743/13348260/4ad4b9f0-dc27-11e5-9260-e1b25c81bcd7.png)
![screen shot 2016-02-26 at 1 18 46 am](https://cloud.githubusercontent.com/assets/1131743/13348257/45ed8f7a-dc27-11e5-903d-a8d7db052cb3.png)

My latest version no longer forces transparent and filled blocks to have exactly the same spacing, so we can optimize them independently. (Apart from the demo presentation, it is unlikely that filled and transparent block environments will appear on the same slide.)

![screen shot 2016-02-26 at 1 19 32 am](https://cloud.githubusercontent.com/assets/1131743/13348266/50313ea0-dc27-11e5-84be-84441cd9c646.png)
![screen shot 2016-02-26 at 1 19 19 am](https://cloud.githubusercontent.com/assets/1131743/13348264/4e6ae062-dc27-11e5-9624-92bb512714eb.png)

I would appreciate you two giving this a closer look and sharing your opinions and suggestions!

benjamin-weiss commented:
I still don't like the behavior of the `beamercolorbox` `sep` parameter, but with this small value it doesn't really stand out. So although I would prefer an implementation similar to my frame title struts I'm fine with it here. Overall it is a big improvement to the previous implementation, so I would just merge it right now. The only thing you maybe could do is to reduce the number of hardcoded lengths.

But we maybe should get back to this issue at some point and try to figure out a real solution or better workaround for the `ifbeamercolorempty` bug. But right know I also don't see a better way.

PS: Just a minor issue as probably hardly anybody is ever gonna do it, but setting only the body background to empty also looks strange.
```TeX
\documentclass{beamer}
\usetheme[block=fill]{metropolis}
\begin{document}
\begin{frame}{Blocks}
  \setbeamercolor{block body}{bg=}
  \begin{block}{Default}
     Block content.
  \end{block}
\end{frame}
\end{document}
```

rchurchley commented:
> The only thing you maybe could do is to reduce the number of hardcoded lengths.

I've refactored the code again to make the lengths less hacky. and added comments so it's a bit clearer what's going on! I don't see any regressions on the presentation I've been using to test, but I'd appreciate one last scan by you guys and then we can merge it.

> Just a minor issue as probably hardly anybody is ever gonna do it, but setting only the body background to empty also looks strange.

I think that's probably unavoidable. Fortunately, anybody who wanted to do that for some reason could instead always use `\setbeamercolor{block body}{bg=normal text.bg}` to have a transparent-looking background with fill-compatible spacing.

benjamin-weiss commented:
On last thing. You could discard your changes to the outer theme that our two PRs do not have any conflicts.

rchurchley commented:
Good point, and now done. I think this should be ready to merge now?

matze commented:
Sorry for the late reply, have been on a short vacation. I think this and #183 are ready for a merge. Thanks!
